torch/csrc/autograd/engine.cpp

```diff
@@ -736,12 +748,33 @@ auto Engine::start_threads() -> void {
   ready_queues_ = std::vector<std::shared_ptr<ReadyQueue>>(num_threads);
   for (auto& queue : ready_queues_)
     queue.reset(new ReadyQueue());
+
+  thread_pool_shared_ = std::make_shared<ThreadPoolShared>();
+
   for (int i = 0; i < num_threads; ++i) {
     std::thread t(&Engine::thread_init, this, i - 1);
     t.detach();
   }
 }
 
+void Engine::add_thread_pool_task(GraphTask *graph_task) {
+  std::unique_lock<std::mutex> lck(thread_pool_shared_->mutex_);
+  // There may already be some items on the graphtasks_queue_ added by other
+  // threads but not enough workers to get to the the new task that will be
+  // added
+  bool create_thread = (thread_pool_shared_->num_workers_ <= thread_pool_shared_->graphtasks_queue_.size());
+  thread_pool_shared_->graphtasks_queue_.push(graph_task);
+  // Don't need to be holding the lock while actually creating the thread
+  lck.unlock();
+  if (create_thread) {
+    std::thread t(&Engine::reentrant_thread_init, this);
```

Comment:
@ilia-cher do we have recommended way to create threads?

Reply:
oh, we definitely don't want to just create std::thread's..

Reply:
To be clear, the autograd engine has always maintained its own little thread pool, of one thread per CPU / CUDA device, even prior to this patch. If you want to say that autograd should actually use the existing thread pool implementation, that may be a reasonable stance to take, but (1) it's a definite expansion of scope beyond what this PR is trying to do (which is to fix #6959) and (2) autograd has some very particular behavioral requirements for its thread pool (specifically, tasks must be associated with a *specific* worker thread on the pool) and it is not obvious to me that the generic thread pool implementation satisfies them (I assume you are referring to `c10/core/thread_pool.h`)